### ISSUE TITLE
systemd initrd loads graphics driver too late

### ISSUE THREAD

**Issue by ncfavier** [NixOS/nixpkgs#202846]
The systemd-based initrd parallelises the early boot process, which is good except that it removes the [ordering](https://github.com/NixOS/nixpkgs/blob/cdcd41104a3eee16aad805dda72f374d85048e86/nixos/modules/system/boot/stage-1-init.sh#L233-L259) between loading kernel modules (`systemd-modules-load.service`) and udev (`systemd-udevd.service`). This causes issues for me because my graphics driver (`amdgpu`) takes about 3 seconds to load, at which point:

1. the console font set by the `90-vconsole.rules` udev rule is reset to the default, and only reset correctly much later, when `reload-systemd-vconsole-setup.service` is pulled by `multi-user.target`;
2. as reported [here](https://github.com/NixOS/nixpkgs/issues/178345#issuecomment-1205053816), the console gets scrolled up a few lines while the LUKS password prompt stays floating in the middle of the screen. I don't know if this is a bug with amdgpu, Linux, systemd, or how to even begin troubleshooting it.

Setting

```nix
{ boot.initrd.systemd.services.systemd-udevd.after = [ "systemd-modules-load.service" ]; }
```

fixes both issues.

If this turns out to be the correct solution to this problem then it should be taken upstream, but I wanted to ask opinions here first.

There is also the alternative of considering issues 1 and 2 as *bugs* (with Linux? amdgpu?) and trying to fix them, which might take a long time.

cc @oxalica @ElvishJerricco @NickCao @lheckemann @dasJ

**Comment by NickCao** [comment]
Shouldn't udev be well prepared to handle these module load events, as it is the **Dynamic** device management daemon after all, maybe just tweaking the udev rule a bit to trigger `90-vconsole.rules` again after the module load is the right fix. Still I'm curious about the design behind all these: why we need **both** a service and a udev rule to get the console right.

**Comment by ncfavier** [comment]
> maybe just tweaking the udev rule a bit to trigger `90-vconsole.rules` again after the module load is the right fix

That would be great, but I can't figure it out: looking at udev debug logs, the only event emitted for `vtcon0` is the initial `add`. I tried adding the same rule for the `amdgpu` device and for `fb*` devices, but in both cases `systemd-vconsole-setup` fails.

> why we need **both** a service and a udev rule to get the console right

I think the udev rule is for initialising the vconsole as soon as it appears, while the service is for reconfiguring it when changes are made to vconsole.conf.

**Comment by NickCao** [comment]
Searching through the issues led me to this: https://github.com/systemd/systemd/issues/2612

> Well, I don't see how this could ever work: we simply don't know whether there will be another KMS driver showing up or not. Device probing is full async, hence it might appear any time, and there's no point in time where we know everything has shown up. Hence wey cannot delay vconsole accordingly.

However it also mentions:

> in systemd 231-232 we reworked the vconsole font setup process, fixing various issues

While not pointing out what the reworked process looks like,

**Comment by ncfavier** [comment]
I saw that, it's rather old, I'm pretty sure the rework is https://github.com/systemd/systemd/pull/3742, which is just the current state of things as we know it.

**Comment by NickCao** [comment]
If that's the rework, it does not seem to address the race conditions, or does it?

**Comment by ncfavier** [comment]
At least not the one we care about now.

**Comment by NickCao** [comment]
Per the information I have had, the vconsole setup issue is a won't fix, due to the natural of async device probing, imaging the scenario that a particular mother board powering up the GPU halfway through the boot process, after `systemd-modules-load.service`, whatever ordering of services won't be able to handle this. The mysterious scrolling requires further investigation though.

**Comment by NickCao** [comment]
If there has to be a fix, I think it's the kernel VT infrastructure's responsibility to either persist the vconsole states across the driver load, or inform udev of the change in order to trigger a reload.

**Comment by ncfavier** [comment]
> sending a message to the kernel mailing list

Do you want to take care of it? You seem more competent than I am.

**Comment by 9ary** [comment]
I would, but I'm very overwhelmed with other things right now so I'm not really sure that I can give this proper attention.

**Comment by rikkaneko** [comment]
I found that the systemd-vconsole-setup is called by the udev rules only but does not include the systemd-vconsole-setup.service. The solution would be to include this service to the initrd and add the systemd-modules-load.service as a dependency.

**Comment by 9ary** [comment]
That's incorrect as far as I know. The udev rule exists for event-driven initialization and it is working as intended. This is a kernel problem and it would be much more useful to fix the root cause rather than come up with bandaid workarounds.

**Comment by 9ary** [comment]
I actually ended up posting to lkml about this because apparently I'm hyperfixating on this anyway.

https://lore.kernel.org/all/CANnEQ3Ef5-XRSVL=RCBuKKhR0oZF+SO2BSSiBigZOyjMeQ7f_g@mail.gmail.com/

**Comment by 9ary** [comment]
Unfortunately I never got a reply. :( Maybe I should try again.

**Comment by ncfavier** [comment]
Maybe relevant?: https://lore.kernel.org/all/20230227020855.1051605-8-sashal@kernel.org/

**Comment by 9ary** [comment]
Interesting find, I wonder if that commit helps or makes things worse.

**Comment by 9ary** [comment]
It would be helpful to have a minimal VM repro of this (using virtio-gpu maybe?).

Also here's my understanding of what's happening and what needs to change: during boot, we initially see efifb and the first fbcon instance. Then when the GPU driver loads, a new instance of fbcon spawns on the new framebuffer driver, and takes over the existing vtcon. This discards the font, and because a new vtcon isn't being added, no event hits udev.

One way to work around this problem is to pass the `quiet` option to the kernel. In that case, the vtcon isn't initialized until fairly late in the boot process, often enough long after the GPU driver has loaded, so the font doesn't get disrupted.

So what needs to happen is the kernel should emit a "change" event whenever a new backend driver takes over a vtcon.

**Comment by rnhmjoj** [comment]
>  why we need both a service and a udev rule to get the console right.

@NickCao, the full lore is here: https://github.com/NixOS/nixpkgs/commit/9be052921004f45a34af2e7a42bd88078b480502

**Comment by ncfavier** [comment]
Can't reproduce any more, closing along with https://github.com/NixOS/nixpkgs/issues/413128

**Comment by ncfavier** [comment]
Spoke too soon. I just updated my system and this seems to have resurfaced.
